### PR TITLE
Adding planar-graph sampler 

### PR DESCRIPTION
**Pull request by mdcoury**
This adds the planar-graph sampler to the dwave-samplers project.  The new classes are available at dwave/samplers/planar and the new PlanarGraphSampler is from dwave.samplers.planar.sampler.

### CODE REVIEW

**Comment by mdcoury** [comment]
Note that I still have to refine the version requirements for scipy and networkx

File: dwave/samplers/planar/planar.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            _ = x, y, xykey = G.nodes[x]['rotation'][xz]
+            _ = y, z, yzkey = G.nodes[y]['rotation'][(y, x, xykey)]
+            _ = z, x, xzkey = G.nodes[z]['rotation'][(z, y, yzkey)]
+
+            if xz != (x, z, xzkey):
+                return False
+    return True
+
+
+# noinspection PyPep8Naming
```
Comment (arcondello) [inline dwave/samplers/planar/planar.py:161]: I believe these are a pycharm specific command, probably best not to include in the actual package.

File: pyproject.toml [hunk 2]
```diff
@@ -6,6 +6,8 @@ requires = [
     'dimod==0.11.3',
     'numpy==1.19.0;python_version<"3.9"',  # C API for numpy.random
     'oldest-supported-numpy;python_version>="3.9"',
+    'networkx>=2.1',
+    'scipy>=1.9.0',
```
Comment (arcondello) [inline pyproject.toml:10]: For build we want to pin.
```
networkx==2.6.3
scipy==1.7.3;python_version<"3.8"
scipy==1.9.2;python_version>="3.8"
```
I think the above versions will work for our entire python range

File: requirements.txt [hunk 3]
```diff
@@ -1,3 +1,5 @@
+networkx>=2.1
```
Comment (arcondello) [inline requirements.txt:1]: will want to pin, probably copying the versions in pyproject.toml

File: setup.cfg [hunk 4]
```diff
@@ -29,6 +29,7 @@ classifiers =
 packages =
     dwave.samplers
     dwave.samplers.greedy
+    dwave.samplers.planar
     dwave.samplers.random
     dwave.samplers.sa
     dwave.samplers.tabu
```
Comment (arcondello) [inline setup.cfg:35]: In this file, below, you want to add `scipy>=1.1.0` and set networkx's lower bound to whatever version [check_planarity](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.planarity.check_planarity.html#check-planarity) came in.

File: dwave/samplers/planar/sampler.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        """
+
+        :param bqm:
+        :param pos:
+        :param parameters:
+        :return:
+        """
+
+        if pos is None:
+            pos = {}
```
Comment (arcondello) [inline dwave/samplers/planar/sampler.py:36]: I believe this case will currently fail, needs tests

File: dwave/samplers/planar/planar.py (hunk 1 above)
Comment (mdcoury) [reply dwave/samplers/planar/planar.py:161]: Agreed: I'm taking them out as I add the comments/documentation

File: dwave/samplers/planar/sampler.py (hunk 5 above)
Comment (mdcoury) [reply dwave/samplers/planar/sampler.py:36]: It does fail: yes.  I assume that is the expected behaviour?

Comment (arcondello) [reply dwave/samplers/planar/sampler.py:36]: In that case, you should generate the planar embedding with [check_planarity](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.planarity.check_planarity.html#check-planarity). It's in a slightly different format, but shouldn't be too hard to work with both (or translate one into the other)

File: dwave/samplers/planar/__init__.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from dwave.samplers.planar.util import *
```
Comment (arcondello) [inline dwave/samplers/planar/__init__.py:1]: I know it's silly, but this needs a copyright header as well

File: dwave/samplers/planar/__init__.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+import dwave.samplers.planar.util
+
+from dwave.samplers.planar.planar import *
```
Comment (arcondello) [inline dwave/samplers/planar/__init__.py:4]: I would replace this whole file with just
```python
<header>

from dwave.samplers.planar.sampler import *
```
You don't need the rest of it

**Comment by arcondello** [comment]
Should add `from dwave.samplers.planar import *` to `dwave/samplers/__init__.py`

File: dwave/samplers/planar/sampler.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        return ret
+
+
+def _determine_pos(G: MultiGraph) -> dict:
+    is_planar, P = networkx.check_planarity(G)
+    if not is_planar:
+        raise ValueError("The provided BQM does not yield a planar embedding")
+
+    layout = nx.planar_layout(P)
+    return {k: tuple(v) for k, v in layout.items()}
```
[the remark this answers is not in the document]
Comment (mdcoury) [reply dwave/samplers/planar/sampler.py]: You're right... for some reason I thought that wasn't working yesterday

**Review by arcondello — APPROVED** [review]

**Pull request merged**